# Module 06 Lab - Regression and Classification Models**Objective:** To understand the difference between regression and classification and to build your first linear models for both tasks.**In this lab, you will write the code to train and evaluate the models.**

## Part 1: Regression vs. ClassificationThis is the most fundamental distinction between types of supervised learning problems.*   **Regression:** The goal is to predict a **continuous numerical value**.     *   *Examples:* Predicting the price of a house, the temperature tomorrow, or the stock price.*   **Classification:** The goal is to predict a **discrete category or class label**.    *   *Examples:* Predicting if an email is spam or not spam, if a flower is a setosa, versicolor, or virginica, or if a customer will churn or not.**In this lab, we will tackle one of each.**

## Part 2: Linear Regression**Concept:** Linear Regression is used to predict a continuous value. It works by finding the best-fitting straight line through the data points. The model learns a "slope" (coefficient) for each feature and an "intercept".*   **Problem:** We will predict the `Fare` of a Titanic passenger based on their `Age` and `Pclass`.

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
# Load and prepare the datadf = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
# For simplicity, we'll drop rows with missing agedf.dropna(subset=['Age'], inplace=True)
# Define features and targetfeatures = ['Age', 'Pclass']target_reg = 'Fare'X_reg = df[features]y_reg = df[target_reg]
# Split the dataX_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=42)

### Task 1: Train and Evaluate a Linear Regression Model**Your Task:**1.  Create an instance of the `LinearRegression` model.2.  Train the model using the training data (`X_train_reg`, `y_train_reg`).3.  Make predictions on the test data.4.  Evaluate the model using `mean_squared_error`. This metric tells us the average of the squared differences between the predicted and actual values.

In [11]:
# Load and prepare the data
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

# Drop rows with missing Age
df.dropna(subset=['Age'], inplace=True)

# Define features and target
features = ['Age', 'Pclass']
target_reg = 'Fare'

X_reg = df[features]
y_reg = df[target_reg]

# Split the data
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)

## Part 3: Logistic Regression**Concept:** Despite its name, Logistic Regression is a **classification** algorithm. It works by calculating the probability that a given input belongs to a certain class. It's one of the most widely used and interpretable classification models.*   **Problem:** We will predict whether a passenger `Survived` based on their `Age`, `Pclass`, and `Sex`.

In [7]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Encode Sex column
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})

# Define features and target
features_cls = ['Age', 'Pclass', 'Sex']
target_cls = 'Survived'

X_cls = df[features_cls]
y_cls = df[target_cls]

# Split the data
X_train_cls, X_test_cls, y_train_cls, y_test_cls = train_test_split(
    X_cls, y_cls, test_size=0.2, random_state=42
)

### Task 2: Train and Evaluate a Logistic Regression Model**Your Task:**1.  Create an instance of the `LogisticRegression` model.2.  Train the model using the classification training data.3.  Make predictions on the test data.4.  Evaluate the model using `accuracy_score`.

In [8]:
# 1. Create the model instance
log_model = LogisticRegression()

# 2. Train the model
log_model.fit(X_train_cls, y_train_cls)

# 3. Make predictions
y_pred_cls = log_model.predict(X_test_cls)

# 4. Evaluate the model
accuracy = accuracy_score(y_test_cls, y_pred_cls)

print(f"Accuracy for Survival Prediction: {accuracy:.2%}")

Accuracy for Survival Prediction: 74.83%


In [12]:
lr_model = LinearRegression()
lr_model.fit(X_train_reg, y_train_reg)
print(lr_model.coef_)

[ -0.43695889 -38.29728935]


## 📝 Knowledge Check**Instructions:** Answer the following questions in this markdown cell.
1.  **In your own words, what is the key difference between a regression problem and a classification problem?
2.  **The `LinearRegression` model has an attribute called `.coef_`. After you train the model, print `lr_model.coef_`. What do these numbers represent?
3.  **Why did we use `mean_squared_error` to evaluate the regression model but `accuracy_score` for the classification model?** Why wouldn't accuracy be a good metric for the fare prediction task?

**[ENTER YOUR ANSWERS HERE]**

1. The main difference is what kind of result the model is trying to predict. A regression problem predicts a continuous numerical value, like the fare of a Titanic passenger or the price of a house. A classification problem predicts a category or label, like whether a passenger survived or not. Basically, regression answers “how much?” while classification answers “which group?”

2. The .coef_ values represent how much each feature affects the predicted fare. In my model, the first coefficient (-0.44) corresponds to Age, and the second coefficient (-38.30) corresponds to Pclass. This means that for every 1-year increase in age, the predicted fare decreases by about 0.44. For every 1-level increase in passenger class, the predicted fare decreases by about $38.30. This shows that passenger class has a much bigger impact on fare than age.

3. We used mean squared error (MSE) for the regression model because the goal was to predict Fare, which is a continuous numerical value. MSE measures how far the predicted fares are from the actual fares by calculating the average squared difference. This helps show how close the model’s predictions are.

We used accuracy score for the classification model because the goal was to predict Survived, which is a category (0 or 1). Accuracy measures how often the model correctly predicts the right class.

Accuracy would not be a good metric for fare prediction because fares can have many possible values. Even if the model predicts a fare very close to the actual amount, it would still count as wrong if it is not exact. MSE is better because it measures the size of the error instead of requiring an exact match.